pyfr/plugins/base.py

```diff
@@ -161,3 +165,87 @@ def _prepare_region_data_eset(self, intg, eset):
 
             self._ele_regions.append((doff, etype, darr))
             self._ele_region_data[f'{etype}_idxs'] = darr
+
+    def prepare_surface_data(self, intg):
+        if self._rgn:
+            self._surf_region_data = self._rgn.surface_faces(intg.system.mesh,
+                                                             intg.rallocs)
+
+class ElementMixin:
+    def __init__(self, intg, *args, **kwargs):
+        super().__init__(intg, *args, **kwargs)
+
+        self._surf_quad = {}
+
+    def surf_quad(self, itype, proj, *args):
+        try:
+            ppts, pwts = self._surf_quad[itype, proj]
+        except KeyError:
+            # Obtain quadrature info
+            rname = self.cfg.get(f'solver-interfaces-{itype}', 'flux-pts')
+
+            # Quadrature rule (default to that of the solution points)
+            qrule = self.cfg.get(self.cfgsect, f'quad-pts-{itype}', rname)
+            try:
+                qdeg = self.cfg.getint(self.cfgsect, f'quad-deg-{itype}')
+            except NoOptionError:
+                qdeg = self.cfg.getint(self.cfgsect, 'quad-deg')
+
+            q = get_quadrule(itype, qrule, qdeg=qdeg, flags='s')
+            ppts, pwts = _proj_pts(proj, q.pts), q.wts
+            self._surf_quad[itype, proj] = (ppts, pwts)
+        return ppts, pwts
+
+    @staticmethod
+    def _plocs(eles, eidx, pts):
+        nspts, neles, ndims = eles.nspts, len(eidx), eles.ndims
+
+        # Construct the physical location operator matrix
+        plocop = eles.basis.sbasis.nodal_basis_at(pts)
+
+        # Apply the operator to the mesh elements and reshape
+        e_surf = eles.eles[:,eidx]
+        ploc = plocop @ e_surf.reshape(nspts, -1)
+        ploc = ploc.reshape(len(pts), neles, ndims)
+
+        return ploc
+
+    @staticmethod
+    def _jacs(eles, eidx, pts):
+        neles, ndims = eles.neles, eles.ndims
+
+        smats, djacs = eles._smats_djacs_mpts
+        djacs = djacs[:,eidx]
+        smats = smats[...,[e + neles*i for i in range(ndims) for e in eidx]]
+
+        # Interpoation shape operator
+        m0 = eles.basis.mbasis.nodal_basis_at(pts)
+
+        # Interpolate the djacs
+        djac = m0 @ djacs
+
+        # Interpolate the smats
+        smats = np.array([m0 @ smat for smat in smats])
+
+        return smats.reshape(ndims, -1, ndims, len(eidx)), 1.0 / djac
+
+    @staticmethod
+    def _pnorm(smats, norm):
+        npts = smats.shape[1]
+        smat = smats.transpose(1, 3, 0, 2)
+
+        norms = np.vstack([[fn]*n for fn, n in zip([norm], [npts])])
```

Comment:
Not sure what is going on here with zip, but it seems broken.

Reply:
This is taken from the norm code elsewhere. It just makes an array of norm vectors